### PR TITLE
Zhijxu/fix softmax fp16

### PR DESCRIPTION
**Pull request by zhijxu-MS**
in fp16 input, the softmax will return nan in some case, 

the reason is because in float16 dtype, std::numeric_limits<float16>::infinity() will return 0 instead of inf

### CODE REVIEW

**Review by pengwa — APPROVED** [review]
Thanks for the fix.

File: onnxruntime/core/providers/cuda/math/softmax_warpwise_impl.cuh [hunk 1]
```diff
@@ -176,17 +174,18 @@ __global__ void softmax_warp_forward_resource_efficient(output_t* dst, const inp
...
-  input_t (&elements)[WARP_ITERATIONS][WARP_SIZE] = *reinterpret_cast<input_t (*)[WARP_ITERATIONS][WARP_SIZE]>(smem);
+  input_t(&elements)[WARP_ITERATIONS][WARP_SIZE] = *reinterpret_cast<input_t(*)[WARP_ITERATIONS][WARP_SIZE]>(smem);
 #pragma unroll
   for (int it = 0; it < WARP_ITERATIONS; ++it) {
     int element_index = local_idx + it * WARP_SIZE;
     if (element_index < element_count) {
       elements[it][local_idx] = src[it * WARP_SIZE];
     } else {
-      elements[it][local_idx] = -std::numeric_limits<input_t>::infinity();
+      static_assert(!std::is_same<acc_t, half>::value, "acc_t can no be half, as the infinity function will return 0 instead of inf");
```
Comment (tianleiwu) [inline onnxruntime/core/providers/cuda/math/softmax_warpwise_impl.cuh:187]: static_assert(std::numeric_limits<acc_t>::has_infinity());

File: onnxruntime/core/providers/cuda/math/softmax_warpwise_impl.cuh [hunk 2]
```diff
@@ -176,17 +174,18 @@ __global__ void softmax_warp_forward_resource_efficient(output_t* dst, const inp
...
+      elements[it][local_idx] = (input_t)-std::numeric_limits<acc_t>::infinity();
```
Comment (tianleiwu) [inline onnxruntime/core/providers/cuda/math/softmax_warpwise_impl.cuh:188]: nit: static_cast<input_t>(-std::numeric_limits<acc_t>::infinity())
nit: Add `#include <limits>`

Comment (zhijxu-MS) [reply onnxruntime/core/providers/cuda/math/softmax_warpwise_impl.cuh:188]: "limits" is included in core/providers/cuda/cu_inc/common.cuh already, thanks

File: onnxruntime/core/providers/cuda/math/softmax_warpwise_impl.cuh [hunk 3]
```diff
@@ -176,17 +174,19 @@ __global__ void softmax_warp_forward_resource_efficient(output_t* dst, const inp
...
-  input_t (&elements)[WARP_ITERATIONS][WARP_SIZE] = *reinterpret_cast<input_t (*)[WARP_ITERATIONS][WARP_SIZE]>(smem);
+  input_t(&elements)[WARP_ITERATIONS][WARP_SIZE] = *reinterpret_cast<input_t(*)[WARP_ITERATIONS][WARP_SIZE]>(smem);
 #pragma unroll
   for (int it = 0; it < WARP_ITERATIONS; ++it) {
     int element_index = local_idx + it * WARP_SIZE;
     if (element_index < element_count) {
       elements[it][local_idx] = src[it * WARP_SIZE];
     } else {
-      elements[it][local_idx] = -std::numeric_limits<input_t>::infinity();
+      static_assert(std::numeric_limits<acc_t>::has_infinity,
```
Comment (tianleiwu) [inline onnxruntime/core/providers/cuda/math/softmax_warpwise_impl.cuh:187]: has_infinity()

**Review by tianleiwu — APPROVED** [review]

**Review by pengwa — APPROVED** [review]

**Pull request merged**